### ISSUE TITLE
Batch Update of SEO Scores

### ISSUE THREAD

**Issue by gordonbanderson** [micschk/silverstripe-liveseo#7]
Currently if the SEO scoring mechanism changes then one would have to load and save each page individually as the score is calculated by JavaScript.  It would be good if there was some way, e.g. a development task, to update these scores in a batch like manner.

The problem here though is that the score is calculated using JavaScript which is not available to command line PHP without installing likes of Node, the only other option I can think of is pushing the score calculation back to PHP and calling a suitable method using AJAX.  However this would slow things down in the interface.

Just a placeholder for now, it's a nice to have.

**Comment by micschk** [comment]
I agree, and have been meaning to refactor the whole scoring proces to make it:
- more configurable
- more robust
- less coupled

What I've come up with as an approach, would roughly be:
- Have the javascript generated from php (templated js, like how DisplayLogic works), this way, knowledge about the calculations would have to be moved to - and be available in php (and configuration to yaml)
- Generalize seo tests that can be applied to fields/properties, e.g. SEOTest_CountTags(htmltag), _Count(words/chars), _CountOccurences(subject), _CheckAttributes(subject) (I think this list about covers the current tests, testing attributes may need a bit of work)
- By generalizing these tests, we can make them both know how to calculate their score themselves, as well as how to generate themed js to live-calculate the score
- These tests could be configured individually (config.yml), but we may also look at configuring them to cover all fields of a certain type (like Fluent gets its translatable fields)
- A generic _AjaxCheck(subject) could leave room to have a kind of fallback for more complex calculations, for example check use of the current page subject on other pages (not sure of its use though)

Curious to get your opinion on this @gordonbanderson?

**Comment by micschk** [comment]
Just a quick note regarding synonyms, we could look at:
- http://thesaurus.altervista.org/ (synonyms api it_IT, fr_FR, de_DE, en_US, el_GR, es_ES, de_DE, no_NO, pt_PT, ro_RO, ru_RU, sk_SK)
- https://words.bighugelabs.com/ or https://www.wordsapi.com/ (en only syn apis)
- or add some kind of custom synonyms getter

**Comment by gordonbanderson** [comment]
re 'DisplayLogic' can you please clarify, lots of different hits for that phrase in search engines.

Interesting idea re using third party services for synonyms, Words API looks nice.  Guess these would have to be cached though :)  I came across an openly licensed thesaurus file when I was working on a module for Elasticsearch, will see if I can find it again.

**Comment by micschk** [comment]
DisplayLogic: https://github.com/unclecheese/silverstripe-display-logic

From a quick glance, I don't think it uses themed js under the hood. But the point was regarding the syntax to generate js behaviour, e.g.;

```
$products->displayIf("HasProducts")->isChecked();
```

**Comment by gordonbanderson** [comment]
I had a brief look at the tools you mentioned.  I've implemented a very rough import using the  BigHugeLabs just to get a feel of what's available.  It not only offers synonyms but also antonyms (opposite meaning) and similar.  One thing I did not think about initially was synonyms based on the classification of a word, e.g. 'rain' as a noun or 'rain' as a verb. Curious to know how this would be represented in a GUI interface.

I think synonyms needs to be stored locally to avoid hits on third party services, and the results of third party services need normalized so that they can be stored consistently in SilverStripe.  This sounds to me like a separate module.  This would be advantageous for several reasons:
- Caching as mentioned above
- Editing of synonyms by a content admin, something that may be required if the content is in a niche area English language wise such as a biological science.
- Import from a third party file.  One such example of a suitable source http://www.gutenberg.org/ebooks/10681, although the text file does not appear to indicate verb, adjective etc
